hkmc2/shared/src/main/scala/hkmc2/codegen/Block.scala

```diff
@@ -52,14 +54,147 @@ sealed abstract class Block extends Product with AutoLocated:
     case TryBlock(sub, fin, rst) => 1 + sub.size + fin.size + rst.size
     case Label(_, bod, rst) => 1 + bod.size + rst.size
   
+  // ignoring blocks inside functions and handle block
+  def map(f: Block => Block): Block = this match
+    case _: Return | _: Throw | _: End | _: Break | _: Continue | _: HandleBlockReturn => this
+    case Match(scrut, arms, dflt, rst) => Match(scrut, arms.map(_ -> f(_)), dflt.map(f), f(rst))
+    case Label(lbl, bod, rst) => Label(lbl, f(bod), f(rst))
+    case Begin(sub, rst) => Begin(f(sub), f(rst))
+    case TryBlock(sub, fin, rst) => TryBlock(f(sub), f(fin), f(rst))
+    case Assign(l, r, rst) => Assign(l, r, f(rst))
+    case b @ AssignField(l, n, r, rst) => AssignField(l, n, r, f(rst))(b.symbol)
+    case Define(defn, rst) => Define(defn, f(rst))
+    case HandleBlock(l, res, par, cls, hdr, bod, rst) => HandleBlock(l, res, par, cls, hdr, bod, f(rst))
+  
+  def mapResult(f: Result => Opt[(Result => Block) => Block]): Block = this match
+    case Return(res, implct) => f(res).map(_(Return(_, implct))).getOrElse(this)
+    case Throw(exc) => f(exc).map(_(Throw(_))).getOrElse(this)
+    case Assign(l, r, rst) => f(r).map(_(Assign(l, _, rst))).getOrElse(this)
+    case b @ AssignField(l, n, r, rst) => f(r).map(_(AssignField(l, n, _, rst)(b.symbol))).getOrElse(this)
+    case HandleBlockReturn(res) => f(res).map(_(HandleBlockReturn(_))).getOrElse(this)
+    case _: End | _: Break | _: Continue | _: Match | _: Label | _: Begin | _: TryBlock | _: Define | _: HandleBlock | _: HandleBlockReturn => this
+  
+  def mapPath(f: Path => Path): Block = this match
+    case Return(res: Path, implct) => Return(f(res), implct)
+    case Throw(exc: Path) => Throw(f(exc))
+    case Assign(l, r: Path, rst) => Assign(l, f(r), rst)
+    case b @ AssignField(l, n, r: Path, rst) => AssignField(l, n, f(r), rst)(b.symbol)
+    case Match(scrut: Path, arms, dflt, rst) => Match(f(scrut), arms, dflt, rst)
+    case Define(ValDefn(owner, k, sym, rhs: Path), rst) => Define(ValDefn(owner, k, sym, f(rhs)), rst)
+    case HandleBlockReturn(res: Path) => HandleBlockReturn(f(res))
+    case _: Return | _: Throw | _: Assign | _: AssignField | _: End | _: Break | _: Continue | _: Match | _: Label | _: Begin | _: TryBlock | _: Define | _: HandleBlock | _: HandleBlockReturn => this.mapResult {
+      case r @ Call(fun, args) => S(_(Call(f(fun), args.map(arg => Arg(arg.spread, f(arg.value))))(r.isMlsFun)))
+      case r @ Instantiate(cls, args) => S(_(Instantiate(f(cls), args.map(f))))
+      case r => N
+    }
+  
+  def mapSyms(f: Local => Local)(using State): Block = 
+    def g(t: Local) = t match 
+      case t: TermSymbol =>
+        TermSymbol(t.k, t.owner.map(symMap), t.id)
+  
+      case _ => f(t)
+    def symMap[T <: Local : ClassTag](l: T): T = g(l) match
+      case r : T => r
```

Comment:
The use of `ClassTag` for pattern matching is a completely unsound misfeature of Scala 2. Doesn't this yield a warning by now?

The proper way in Scala 3 is through [TypeTest](https://docs.scala-lang.org/scala3/reference/other-new-features/type-test.html).

Though I'm not convinced `symMap` is a very justified function to have. It looks like it's a hack fixing a symptom of another problem. I'm pretty sure the design of the current `Symbol` class hierarchy has problems and could be improved and would rather try to understand what the root problem being addressed here is.

Reply:
The root problem is that the input of `mapSyms` *must* be a function that maps symbols of the same subtype to the same subtype because many blocks' parameters require specific types of symbols, and these types must be preserved after being mapped. But, I'm not sure how to express such a function using the type system other than this kind of hack.

Reply:
Here is how I suggest to fix this:
* You define a `SymbolSubst` class that contains one Map per type of symbol to substitute.
* You declare a generic `subst(using SymbolSubst): Symbol` method in `Symbol` and give implementations in the Symbol subclasses  that _refine_ this signature, e.g. `subst(using BlockMemberSymbol): BlockMemberSymbol` in `BlockMemberSymbol`.

Note that you don't need to define one such Map + refining implementation per concrete leaf subclass of Symbol. For some symbol sub-hierarchies, it will be sufficient to define a single one.